<a href="https://colab.research.google.com/github/erasmomendez/03MIARAlgoritmos/blob/main/Proyecto_G11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Actividad - Proyecto práctico


> La actividad se desarrollará en grupos pre-definidos de 4 alumnos. Se debe indicar los nombres en orden alfabético (de apellidos). Recordad que esta actividad se corresponde con un 30% de la nota final de la asignatura. Se debe entregar entregar el trabajo en la presente notebook.

*   Alumno 1: Nicolás Manzano Carretero
*   Alumno 2: Javier Martínez Arrieta
*   Alumno 3: Cristian Martos García
*   Alumno 4: Erasmo Oscar Méndez Chávez






---
## **PARTE 1** - Instalación y requisitos previos

> Las prácticas han sido preparadas para poder realizarse en el entorno de trabajo de Google Colab. Sin embargo, esta plataforma presenta ciertas incompatibilidades a la hora de visualizar la renderización en gym. Por ello, para obtener estas visualizaciones, se deberá trasladar el entorno de trabajo a local. Por ello, el presente dosier presenta instrucciones para poder trabajar en ambos entornos. Siga los siguientes pasos para un correcto funcionamiento:
1.   **LOCAL:** Preparar el enviroment, siguiendo las intrucciones detalladas en la sección *1.1.Preparar enviroment*.
2.  **AMBOS:** Modificar las variables "mount" y "drive_mount" a la carpeta de trabajo en drive en el caso de estar en Colab, y ejecturar la celda *1.2.Localizar entorno de trabajo*.
3. **COLAB:** se deberá ejecutar las celdas correspondientes al montaje de la carpeta de trabajo en Drive. Esta corresponde a la sección *1.3.Montar carpeta de datos local*.
4.  **AMBOS:** Instalar las librerías necesarias, siguiendo la sección *1.4.Instalar librerías necesarias*.


---
### 1.1. Preparar enviroment (solo local)



> Para preparar el entorno de trabajo en local, se han seguido los siguientes pasos:
1. En Windows, puede ser necesario instalar las C++ Build Tools. Para ello, siga los siguientes pasos: https://towardsdatascience.com/how-to-install-openai-gym-in-a-windows-environment-338969e24d30.
2. Instalar Anaconda
3. Siguiendo el código que se presenta comentado en la próxima celda: Crear un enviroment, cambiar la ruta de trabajo, e instalar librerías básicas.


```
conda create --name miar_rl python=3.8
conda activate miar_rl
cd "PATH_TO_FOLDER"
conda install git
pip install jupyter
```


4. Abrir la notebook con *jupyter-notebook*.



```
jupyter-notebook
```


---
### 1.2. Localizar entorno de trabajo: Google colab o local

In [ ]:
# ATENCIÓN!! Modificar ruta relativa a la práctica si es distinta (drive_root)
mount='/content/gdrive'
drive_root = mount + "/My Drive/08_MIAR/actividades/proyecto practico"

try:
  from google.colab import drive
  IN_COLAB=True
except:
  IN_COLAB=False

---
### 1.3. Montar carpeta de datos local (solo Colab)

In [ ]:
# Switch to the directory on the Google Drive that you want to use
import os
if IN_COLAB:
  print("We're running Colab")

  if IN_COLAB:
    # Mount the Google Drive at mount
    print("Colab: mounting Google drive on ", mount)

    drive.mount(mount)

    # Create drive_root if it doesn't exist
    create_drive_root = True
    if create_drive_root:
      print("\nColab: making sure ", drive_root, " exists.")
      os.makedirs(drive_root, exist_ok=True)

    # Change to the directory
    print("\nColab: Changing directory to ", drive_root)
    %cd $drive_root
# Verify we're in the correct working directory
%pwd
print("Archivos en el directorio: ")
print(os.listdir())

We're running Colab
Colab: mounting Google drive on  /content/gdrive
Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).

Colab: making sure  /content/gdrive/My Drive/08_MIAR/actividades/proyecto practico  exists.

Colab: Changing directory to  /content/gdrive/My Drive/08_MIAR/actividades/proyecto practico
/content/gdrive/My Drive/08_MIAR/actividades/proyecto practico
Archivos en el directorio: 
[]


---
### 1.4. Instalar librerías necesarias

In [ ]:
if IN_COLAB:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install keras-rl2==1.0.5
  %pip install tensorflow==2.8

  # Añadidos manualmente debido a fallos encontrados
  %pip install protobuf==3.20.0

else:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install pyglet==1.5.0
  %pip install h5py==3.1.0
  %pip install Pillow==9.5.0
  %pip install keras-rl2==1.0.5
  %pip install Keras==2.2.4
  %pip install tensorflow==2.5.3
  %pip install torch==2.0.1
  %pip install agents==1.4.0

  Cloning https://github.com/Kojoley/atari-py.git to /tmp/pip-req-build-cv2ue950
  Running command git clone --filter=blob:none --quiet https://github.com/Kojoley/atari-py.git /tmp/pip-req-build-cv2ue950
  Resolved https://github.com/Kojoley/atari-py.git to commit 86a1e05c0a95e9e6233c3a413521fdb34ca8a089
  Preparing metadata (setup.py) ... done


---
## **PARTE 2**. Enunciado

Consideraciones a tener en cuenta:

- El entorno sobre el que trabajaremos será _SpaceInvaders-v0_ y el algoritmo que usaremos será _DQN_.

- Para nuestro ejercicio, el requisito mínimo será alcanzado cuando el agente consiga una **media de recompensa por encima de 20 puntos en modo test**. Por ello, esta media de la recompensa se calculará a partir del código de test en la última celda del notebook.

Este proyecto práctico consta de tres partes:

1.   Implementar la red neuronal que se usará en la solución
2.   Implementar las distintas piezas de la solución DQN
3.   Justificar la respuesta en relación a los resultados obtenidos

**Rúbrica**: Se valorará la originalidad en la solución aportada, así como la capacidad de discutir los resultados de forma detallada. El requisito mínimo servirá para aprobar la actividad, bajo premisa de que la discusión del resultado sera apropiada.

IMPORTANTE:

* Si no se consigue una puntuación óptima, responder sobre la mejor puntuación obtenida.
* Para entrenamientos largos, recordad que podéis usar checkpoints de vuestros modelos para retomar los entrenamientos. En este caso, recordad cambiar los parámetros adecuadamente (sobre todo los relacionados con el proceso de exploración).
* Se deberá entregar unicamente el notebook y los pesos del mejor modelo en un fichero .zip, de forma organizada.
* Cada alumno deberá de subir la solución de forma individual.

---
## **PARTE 3**. Desarrollo y preguntas

#### Importar librerías

In [ ]:
from __future__ import division

from PIL import Image
import numpy as np
import gym

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation, Flatten, Convolution2D, Permute
from tensorflow.keras.optimizers import Adam
import tensorflow.keras.backend as K

from rl.agents.dqn import DQNAgent
from rl.policy import LinearAnnealedPolicy, BoltzmannQPolicy, EpsGreedyQPolicy
from rl.memory import SequentialMemory
from rl.core import Processor
from rl.callbacks import FileLogger, ModelIntervalCheckpoint

#### Configuración base

In [ ]:
INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4

env_name = 'SpaceInvaders-v0'
env = gym.make(env_name)

np.random.seed(123)
env.seed(123)
nb_actions = env.action_space.n

print("Numero de acciones disponibles:" + str(nb_actions))

print("Formato de las observaciones:")
env.observation_space



Numero de acciones disponibles:6
Formato de las observaciones:


Box(0, 255, (210, 160, 3), uint8)

In [ ]:
class AtariProcessor(Processor):
    def process_observation(self, observation):
        assert observation.ndim == 3  # (height, width, channel)
        img = Image.fromarray(observation)
        img = img.resize(INPUT_SHAPE).convert('L')
        processed_observation = np.array(img)
        assert processed_observation.shape == INPUT_SHAPE
        return processed_observation.astype('uint8')

    def process_state_batch(self, batch):
        processed_batch = batch.astype('float32') / 255.
        return processed_batch

    def process_reward(self, reward):
        return np.clip(reward, -1., 1.)

1. Implementación de la red neuronal

In [ ]:
input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
print(input_shape)

(4, 84, 84)


Construcción del modelo

En la siguiente celda, observamos la función build_model, que  implementa una red neuronal convolucional diseñada para procesar imágenes y tomar decisiones en función de las observaciones del entorno.

1. La primera capa, Permute, reorganiza las dimensiones de entrada para ajustarse al formato esperado por la red convolucional.
2. La red consta de 3 capas convolucionales con funciones de activación ReLU, diseñadas para extraer características visuales. Los filtros se incrementan de 32 a 64 y el tamaño baja de (8,8) a (3,3) al final de las capas convolucionales.
3. Tras aplanar las características extraídas, se añaden 3 capas densas, disminuyendo el tamaño desde 512 hasta el número de acciones posibles.


Una vez el model ha sido creado, se configura la memoria de experiencia, limitada hasta 1000 transiciones en este caso.

A continuación, configuramos la política epsilon-greedy. En este caso, inicialmente el agente elige acciones aleatorias (value_max=1) y gradualmente reduce el valor hasta 0.1, promoviendo más explotación. Además, durante la evaluación, se fija el valor de epsilon en 0.2. Finalmente, la reducción de epsilon ocurre durante 10,000 pasos de entrenamiento.

El último paso sería la configuración del agente. Algunos de los hiperparámetros más importantes serían los siguientes:
-
Gamma = 0.99 --> Factor de descuento que prioriza recompensas futuras (más alto para estrategias a largo plazo)
-
nb_steps_warmup=10000 --> Período de calentamiento donde el agente no actualiza la red Q y explora aleatoriamente.
-
target_model_update=10000 --> Intervalo de actualización para la red objetivo, estabilizando el entrenamiento.
-
train_interval=4 --> El agente actualiza la red Q cada 4 pasos.
-
delta_clip=1.0 --> Restringe los gradientes para mejorar la estabilidad

In [ ]:
def build_model(input_shape, nb_actions):
    model = Sequential()
    model.add(Permute((2, 3, 1), input_shape=(WINDOW_LENGTH,) + input_shape))

    # Más filtros en las capas convolucionales
    model.add(Convolution2D(32, (8, 8), strides=(4, 4), activation='relu'))
    model.add(Convolution2D(64, (4, 4), strides=(2, 2), activation='relu'))
    model.add(Convolution2D(64, (3, 3), activation='relu'))

    model.add(Flatten())

    # Más capacidad en las capas densas
    model.add(Dense(512, activation='relu'))
    model.add(Dense(256, activation='relu'))
    model.add(Dense(nb_actions, activation='linear'))

    print(model.summary())
    return model

# Crear el modelo
model = build_model(INPUT_SHAPE, nb_actions)


# Configuración de la memoria de experiencia
memory = SequentialMemory(limit=1000, window_length=WINDOW_LENGTH)

# Configuración de la política epsilon-greedy
policy = LinearAnnealedPolicy(
    EpsGreedyQPolicy(),
    attr='eps',
    value_max=1.0,
    value_min=0.1,
    value_test=0.2,
    nb_steps=10000
)

dqn = DQNAgent(
    model=model,
    nb_actions=nb_actions,
    policy=policy,
    memory=memory,
    processor=AtariProcessor(),
    nb_steps_warmup=10000,
    gamma=0.99,
    target_model_update=10000,
    train_interval=4,
    delta_clip=1.0
)
dqn.compile(Adam(learning_rate=0.0001), metrics=['mae'])
print("Agente DQN configurado y compilado.")

Model: "sequential_13"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_9 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_33 (Conv2D)          (None, 20, 20, 32)        8224      
                                                                 
 conv2d_34 (Conv2D)          (None, 9, 9, 64)          32832     
                                                                 
 conv2d_35 (Conv2D)          (None, 7, 7, 64)          36928     
                                                                 
 flatten_10 (Flatten)        (None, 3136)              0         
                                                                 
 dense_26 (Dense)            (None, 512)               1606144   
                                                                 
 dense_27 (Dense)            (None, 256)             

En esta celda se configuran callbacks para registrar el progreso del entrenamiento y guardar los pesos del modelo en intervalos regulares, lo cual es fundamental para el seguimiento y recuperación del proceso de entrenamiento.

In [ ]:
# Callbacks para registrar progreso y guardar modelos
weights_filename = 'dqn_space_invaders_weights.h5f'
checkpoint_weights_filename = 'dqn_space_invaders_weights_{step}.h5f'
log_filename = 'dqn_space_invaders_log.json'

callbacks = [
    ModelIntervalCheckpoint(checkpoint_weights_filename, interval=250000),
    FileLogger(log_filename, interval=100)
]

Finalmente, entrenamos el agente. En este caso, solamente se utilizaron 20000 pasos de interacción con el entorno debido a las restricciones de tiempo y de recursos en google colab. Además, una vez finalizado el entrenamiento, se guardaron los pesos finales del modelo.

In [ ]:
dqn.fit(env, nb_steps=20000, visualize=False, verbose=2)

# Guardar los pesos después del entrenamiento
dqn.save_weights('dqn_space_invaders_weights.h5f', overwrite=True)

Training for 20000 steps ...
   635/20000: episode: 1, duration: 2.759s, episode steps: 635, steps per second: 230, episode reward:  8.000, mean reward:  0.013 [ 0.000,  1.000], mean action: 2.504 [0.000, 5.000],  loss: --, mae: --, mean_q: --, mean_eps: --
  1446/20000: episode: 2, duration: 3.459s, episode steps: 811, steps per second: 234, episode reward: 11.000, mean reward:  0.014 [ 0.000,  1.000], mean action: 2.340 [0.000, 5.000],  loss: --, mae: --, mean_q: --, mean_eps: --
  2309/20000: episode: 3, duration: 5.408s, episode steps: 863, steps per second: 160, episode reward: 17.000, mean reward:  0.020 [ 0.000,  1.000], mean action: 2.149 [0.000, 5.000],  loss: --, mae: --, mean_q: --, mean_eps: --
  3285/20000: episode: 4, duration: 4.196s, episode steps: 976, steps per second: 233, episode reward:  7.000, mean reward:  0.007 [ 0.000,  1.000], mean action: 1.926 [0.000, 5.000],  loss: --, mae: --, mean_q: --, mean_eps: --
  4039/20000: episode: 5, duration: 3.274s, episode ste

/usr/local/lib/python3.10/dist-packages/keras/engine/training_v1.py:2079: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,


 10942/20000: episode: 14, duration: 49.911s, episode steps: 1001, steps per second:  20, episode reward: 15.000, mean reward:  0.015 [ 0.000,  1.000], mean action: 2.211 [0.000, 5.000],  loss: 0.005995, mae: 0.022206, mean_q: 0.043855, mean_eps: 0.100000
 12130/20000: episode: 15, duration: 59.820s, episode steps: 1188, steps per second:  20, episode reward: 19.000, mean reward:  0.016 [ 0.000,  1.000], mean action: 2.242 [0.000, 5.000],  loss: 0.006774, mae: 0.023533, mean_q: 0.043884, mean_eps: 0.100000
 12984/20000: episode: 16, duration: 42.073s, episode steps: 854, steps per second:  20, episode reward: 11.000, mean reward:  0.013 [ 0.000,  1.000], mean action: 2.842 [0.000, 5.000],  loss: 0.007841, mae: 0.023015, mean_q: 0.044394, mean_eps: 0.100000
 13342/20000: episode: 17, duration: 17.028s, episode steps: 358, steps per second:  21, episode reward:  5.000, mean reward:  0.014 [ 0.000,  1.000], mean action: 2.573 [0.000, 5.000],  loss: 0.004473, mae: 0.020969, mean_q: 0.03695

**Análisis del entrenamiento realizado:**

*   El agente DQN se entrenó durante 20,000 pasos en el entorno Space Invaders.

*   Esta cantidad de pasos es limitada debido a las restricciones de tiempo y recursos en Google Colab, donde este experimento fue ejecutado.

*   Se completaron 25 episodios durante los 20,000 pasos de entrenamiento.
Al inicio, los episodios fueron cortos (entre 2-6 segundos) debido a la falta de experiencia del agente. Conforme avanza el entrenamiento, los episodios aumentan en duración, llegando a superar 59 segundos, lo que sugiere que el agente está aprendiendo a sobrevivir más tiempo.

*   Las recompensas iniciales fueron bajas, entre 6 y 8 puntos.
Hacia el final del entrenamiento, el agente comenzó a obtener recompensas más altas, alcanzando un máximo de 19 puntos. Esto muestra un progreso en el aprendizaje del agente a medida que interactúa con el entorno.

*   El entrenamiento se realizó de manera eficiente, alcanzando hasta 21 pasos por segundo en algunos episodios.

*   Finalmente, una vez completado el entrenamiento, se procede a evaluar al agente en un entorno de test.

In [ ]:
# Cargar los pesos entrenados (si es necesario)
dqn.load_weights(weights_filename)

# Evaluar el modelo
print("Evaluación en modo test...")
scores = dqn.test(env, nb_episodes=10, visualize=False)

# Verificar la recompensa media
mean_reward = np.mean(scores.history['episode_reward'])
print(f"Recompensa media en modo test: {mean_reward}")

Evaluación en modo test...
Testing for 10 episodes ...
Episode 1: reward: 28.000, steps: 1383
Episode 2: reward: 21.000, steps: 1155
Episode 3: reward: 35.000, steps: 1700
Episode 4: reward: 15.000, steps: 901
Episode 5: reward: 19.000, steps: 763
Episode 6: reward: 24.000, steps: 1400
Episode 7: reward: 31.000, steps: 1673
Episode 8: reward: 13.000, steps: 636
Episode 9: reward: 19.000, steps: 1013
Episode 10: reward: 18.000, steps: 1040
Recompensa media en modo test: 22.3


Después de completar el entrenamiento, se evaluó el agente DQN en 10 episodios para medir su rendimiento sin exploración (modo test). A continuación se detallan los resultados:

1. **Recompensas obtenidas**:
Las recompensas en los episodios varían entre 13 y 35 puntos. El máximo rendimiento se observó en el Episodio 3, con una recompensa de 35 puntos, mientras que el mínimo fue 13 puntos en el Episodio 8.

2. **Promedio de las recompensas**:
La recompensa media obtenida en los 10 episodios fue de 22.3 puntos. Este resultado supera el objetivo requerido de 20 puntos, lo que indica que el agente logró aprender una política efectiva para jugar a Space Invaders.

3. **Pasos por episodio**:
Los episodios más largos en términos de pasos estuvieron alrededor de 1700 pasos (Episodios 3 y 7), lo cual refleja que el agente pudo sobrevivir más tiempo en el juego. Los episodios más cortos, como el Episodio 8 (636 pasos), coinciden con las recompensas más bajas, mostrando que el agente fue derrotado más rápidamente.

3. Justificación de los parámetros seleccionados y de los resultados obtenidos

El agente DQN fue entrenado con el entorno Space Invaders utilizando una red neuronal convolucional optimizada para procesar imágenes de tamaño reducido (84x84 píxeles) y aprovechar información temporal con una ventana de 4 frames. A continuación, se justifican las elecciones de los parámetros más relevantes:

1. **Arquitectura del modelo**:
Se utilizaron tres capas convolucionales con activación ReLU para extraer características visuales relevantes del juego. Se agregaron capas densas de 512 y 256 neuronas para aumentar la capacidad del modelo y permitir un mejor aprendizaje de la política óptima. La capa de salida tiene una activación lineal y un número de neuronas igual al número de acciones disponibles (6 acciones)

2. **Hiperparámetros**:

      - **Gamma** = 0.99 --> Se seleccionó un valor alto para priorizar recompensas futuras, lo cual es crucial en juegos donde las acciones tienen un efecto acumulativo a largo plazo.

      - **Tasa de aprendizaje** = 0.0001 --> Una tasa baja garantiza estabilidad durante el entrenamiento y evita grandes oscilaciones en las actualizaciones de los pesos.

      - **Política epsilon-greedy** --> El parámetro epsilon se redujo linealmente de 1.0 a 0.1 en 10,000 pasos, lo cual permitió un buen equilibrio entre exploración y explotación.

3. **Entrenamiento y resultados**:
El agente fue entrenado durante 20,000 pasos debido a las limitaciones de tiempo y recursos en Google Colab. En modo test, el agente alcanzó una recompensa promedio de 22.3 puntos, superando el objetivo mínimo de 20 puntos.

4. **Memoria de experiencia**:
Se utilizó una memoria de tamaño 1,000 transiciones, lo cual permitió al agente aprender de experiencias pasadas sin depender de observaciones consecutivas.

En conclusión, la combinación de una arquitectura eficiente, hiperparámetros adecuados y una política de exploración controlada permitió al agente superar el objetivo mínimo de recompensa en un entorno limitado. Con más recursos de cómputo, se podrían realizar entrenamientos más extensos para seguir mejorando el rendimiento del agente.


---

#**EXPERIMENTOS SOBRE LOS HIPERPARAMETROS DEL AGENTE DQN O SOBRE EL MODELO**

## Experimento 1

In [ ]:
# Hemos tenido el problema de detencion de la ejecución, debido a que estos modelos necesitaban una gran capacidad de cómputo, más concretamente este modelo necesitaba un numero de steps mínimo de 500k o 1M
# Es por ello que hemos tenido dificultades de ejecucion en colab.
# Por ejemplo, tratamos de reducir el coste computacional para tratar de llegar a los necesarios.

Este es el original el agente DQN

dqn = DQNAgent(
    model=model,
    nb_actions=nb_actions,
    policy=policy,
    memory=memory,
    processor=AtariProcessor(),
    nb_steps_warmup=10000,
    gamma=0.95,
    target_model_update=10000,
    train_interval=10,
    delta_clip=1.0,
    batch_size=64  # Tamaño del batch
)
dqn.compile(Adam(learning_rate=0.0001), metrics=['mae']) que cambios aplicamos para reducir el sotes en el entrenamiento?


# ESTOS SON LOS AJUSTES PARA TRATAR DE REDUCIR EL COSTE COMPUTACIONAL DEL ENTRENAMIENTO.

# Se reduce el tamaño del batch para obtener un entrenamiento más estable, reduciéndolo de 64 a 32
batch_size=32  # Tamaño del batch más pequeño

# Se ajusta el valor del gamma de 95 a 99 ya que un gamma muy bajo podría causar que el agente no aprenda bien a largo plazo.
gamma=0.99

# Se icrementa el valor del step del warmup, esto permitirá que el modelo acumule más experiencia antes de comenzar a entrenar, lo cual puede mejorar la estabilidad.
nb_steps_warmup=20000

# Al reducir el número de pasos entre actualizaciones del modelo objetivo, se consigue evitar que el modelo se sobreajuste demasiado a los datos.
target_model_update=5000  # Actualización más frecuente

# Se ajusta este párametro para entrenar más o menos veces en cada ciclo, lo que puede afectar el ruido en el entrenamiento
train_interval=5  # Entrenar más frecuentemente

# Si el valor de delta_clip es demasiado alto, puede causar actualizaciones de gradientes demasiado grandes, lo que puede generar ruido.
delta_clip=0.5

# Ajustamos la tasa de exploración (epsilon) para equilibrar mejor la exploración y explotación. Si es muy alta, el agente puede estar explorando demasiado y no aprender de manera eficiente.
policy = EpsGreedyQPolicy(epsilon=0.1)

# Otro ajuste fue subir el learning rate
learning_rate=0.0025

In [ ]:
# Política de exploración ajustada para un entrenamiento mas rapido
from rl.policy import LinearAnnealedPolicy, EpsGreedyQPolicy

policy = LinearAnnealedPolicy(
    EpsGreedyQPolicy(),
    attr='eps',
    value_max=1.0,      # Exploración inicial alta
    value_min=0.3,      # Menor exploración más temprano
    value_test=0.05,    # Modo de evaluación con mínima exploración
    nb_steps=200000     # Reducción de pasos para acelerar explotación
)


In [ ]:
# Propuesta de modelo mas complejo, el cual no se pudo ejecutar debido a las limitaciones de hardware en COLAB
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Conv2D, Flatten, Dense, Dropout, BatchNormalization, ReLU

def build_model(input_shape, nb_actions):
    inputs = Input(shape=input_shape)

    # Primera capa convolucional
    x = Conv2D(32, (4, 4), strides=(2, 2), padding='SAME')(inputs)
    x = BatchNormalization()(x)  # Añadimos batch normalization
    x = ReLU()(x)  # ReLU en lugar de LeakyReLU
    x = Dropout(0.3)(x)  # Dropout para la capa convolucional

    # Segunda capa convolucional
    x = Conv2D(64, (4, 4), strides=(2, 2), padding='SAME')(x)
    x = BatchNormalization()(x)
    x = ReLU()(x)
    x = Dropout(0.3)(x)  # Dropout para la capa convolucional

    # Capa convolucional adicional
    x = Conv2D(128, (3, 3), strides=(1, 1), padding='SAME')(x)
    x = BatchNormalization()(x)
    x = ReLU()(x)
    x = Dropout(0.3)(x)  # Dropout para la capa convolucional

    x = Flatten()(x)

    # Capa densa
    x = Dense(128)(x)
    x = ReLU()(x)
    x = Dropout(0.5)(x)  # Dropout para la capa densa

    # Capa de salida
    outputs = Dense(nb_actions, activation='linear')(x)

    model = Model(inputs=inputs, outputs=outputs)
    return model

input_shape = (WINDOW_LENGTH, ) + INPUT_SHAPE
nb_actions = env.action_space.n
model = build_model(input_shape, nb_actions)
print(model.summary())


La arquitectura propuesta es un modelo más complejo diseñado para mejorar el rendimiento del agente DQN en el entorno Space Invaders. Este modelo incorpora varias técnicas avanzadas, como Batch Normalization y Dropout, que permiten estabilizar y generalizar el aprendizaje. Sin embargo, debido a las limitaciones de hardware en Google Colab, no pudo ejecutarse en su totalidad, ya que su mayor cantidad de parámetros y capas incrementa significativamente los requisitos computacionales.

La red comienza con una capa de entrada que recibe los datos con dimensiones (WINDOW_LENGTH, 84, 84), correspondientes a la ventana temporal de 4 frames y a las imágenes redimensionadas del entorno. A continuación, se aplican tres capas convolucionales secuenciales que extraen características visuales progresivamente más complejas. Cada capa utiliza filtros de tamaños decrecientes (4x4 y 3x3), con pasos ajustados para reducir la resolución de manera controlada. Además, se incorpora Batch Normalization después de cada convolución, lo que normaliza la salida de las capas y mejora la estabilidad del entrenamiento.

Para mejorar la generalización del modelo y evitar el sobreajuste, se añade Dropout con tasas del 30% después de cada capa convolucional y del 50% en la capa densa. La capa Flatten convierte las características espaciales extraídas por las convoluciones en un vector unidimensional, preparándolas para su procesamiento en las capas densas. A continuación, se utiliza una capa densa con 128 neuronas, activada con ReLU, que combina las características extraídas y aprende representaciones más abstractas del entorno.

Finalmente, la capa de salida tiene un número de neuronas igual al número de acciones posibles en el entorno. La activación lineal de esta capa permite predecir valores Q continuos para cada acción, fundamentales en el aprendizaje por refuerzo. Esta estructura, con tres niveles de convolución y un procesamiento denso robusto, ofrece una mayor capacidad de aprendizaje en comparación con modelos más simples, ya que captura patrones detallados y abstractos en el entorno visual.

## Experimento 2


En el siguiente experimento se buscó comprobar el funcionamiento para uno de los modelos de red CNN propuestos en el artículo *[Deep Reinforcement Learning to play Space Invaders](https://nihit.github.io/resources/spaceinvaders.pdf)*, escrito por Nihit Desay y Abhimanyu Banerjee. En dicho artículo se puede ver que también realizaron pruebas aplicando Dropout, si bien de acuerdo a los resultados el efecto producido no era el deseado, por lo que se descartó para este experimento. Además, dada las dificultades presentadas para realizar el entrenamiento, se intentó que el modelo a utilizar de entre los propuestos en el artículo no tuviese excesivos parámetros para así conseguir completar el entrenamiento.

In [ ]:
input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
model = Sequential()

# Mostramos el formato de datos de imagen
print('Formato de datos de imagen: ' + K.image_data_format())

# Comprobamos en el formato de imagen si los canales se encuentran al final o al principio y creamos una capa Permute en consecuencia
if K.image_data_format() == 'channels_last':
    # (ancho, alto, canales)
    model.add(Permute((2, 3, 1), input_shape=input_shape))
elif K.image_data_format() == 'channels_first':
    # (canales, ancho, alto)
    model.add(Permute((1, 2, 3), input_shape=input_shape))
else:
    # Mostramos error de no ser ninguna de las anteriores
    raise RuntimeError('image_dim_ordering desconocida.')

# Aplicamos capas convolucionales y reducimos la dimensionalidad
model.add(Convolution2D(32, (8, 8), strides=(4,4), activation='relu'))
model.add(Convolution2D(64, (4, 4), strides=(2,2), activation='relu'))
model.add(Convolution2D(64, (3, 3), strides=(1,1), activation='relu'))

# Aplanamos para conectar con una capa densa
model.add(Flatten())
model.add(Dense(512,activation='relu'))

# La capa de salida contendrá tantas neuronas como posibles acciones tenga el juego, almacenado en la variable nb_actions
model.add(Dense(nb_actions, activation='linear'))
model.save()
print(model.summary())

Formato de datos de imagen: channels_last
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute (Permute)           (None, 84, 84, 4)         0         
                                                                 
 conv2d (Conv2D)             (None, 20, 20, 32)        8224      
                                                                 
 conv2d_1 (Conv2D)           (None, 9, 9, 64)          32832     
                                                                 
 conv2d_2 (Conv2D)           (None, 7, 7, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 3136)              0         
                                                                 
 dense (Dense)               (None, 512)               1606144   
                                                                 
 dense_1 (Dens

El entrenamiento no pudo completarse pero sí se almacenaron checkpoints para ir almacenando los pesos durante el entrenamiento.

In [ ]:
# Siguiendo las recomendaciones, pondremos como límite el mismo número de steps que se establecieron para la política
memory = SequentialMemory(limit=1000000, window_length=WINDOW_LENGTH)


policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), attr='eps',
                              value_max=1., value_min=.1, value_test=.05,
                              nb_steps=1000000)

# Creamos el objeto AtariProcessor
processor = AtariProcessor()

# Definimos el agente
dqn = DQNAgent(model=model, nb_actions=nb_actions, memory=memory,
            processor=processor,
            nb_steps_warmup=50000,       # warmup hace referencia al número de steps de calentamiento
            gamma=0.99,                  # gamma para ponderar la recompensa esperada a futuro en la ecuación de Bellman
            target_model_update=10000,   # Cada cuánto se actualiza la red objetivo, 10000 pasos en este caso
            train_interval=4,            # Cada cuántos stps se entrena. La policy se actualizará cada 4 steps
            policy=policy,
            delta_clip=1.0)

# Realizamos la compilación
dqn.compile(Adam(learning_rate=1e-4, name='Adam'), metrics=['mae'])

# Ejecutamos el entrenamiento
dqn.fit(env, callbacks=callbacks, nb_steps=1250000, log_interval=10000, visualize=False)

In [ ]:
# Hacemos el test con dos de los pesos guardados durante los checkpoints de este otro modelo
# En primer lugar, lo hacemos habiendo hecho 60000 steps, es decir, 10000 después de finalizar el warmup

print('Steps: 60000')
dqn.load_weights('dqn_space_invaders_weights_60000.h5f')
scores = dqn.test(env, nb_episodes=10, visualize=False)
print('Recompensa media: ' + str(np.mean(scores.history['episode_reward'])) + '\n--------------------------------------------')

print('Steps: 490000')
# En segundo lugar, con 490000 steps, los realizados antes de que se parase el entrenamiento
dqn.load_weights('dqn_space_invaders_weights_490000.h5f')
scores = dqn.test(env, nb_episodes=10, visualize=False)
print('Recompensa media: ' + str(np.mean(scores.history['episode_reward'])))

Steps: 60000
Testing for 10 episodes ...
Episode 1: reward: 3.000, steps: 618
Episode 2: reward: 10.000, steps: 969
Episode 3: reward: 3.000, steps: 679
Episode 4: reward: 7.000, steps: 929
Episode 5: reward: 8.000, steps: 980
Episode 6: reward: 3.000, steps: 648
Episode 7: reward: 3.000, steps: 676
Episode 8: reward: 3.000, steps: 683
Episode 9: reward: 3.000, steps: 668
Episode 10: reward: 12.000, steps: 1225
Recompensa media: 5.5
--------------------------------------------
Steps: 490000
Testing for 10 episodes ...
Episode 1: reward: 5.000, steps: 645
Episode 2: reward: 10.000, steps: 786
Episode 3: reward: 7.000, steps: 674
Episode 4: reward: 6.000, steps: 627
Episode 5: reward: 10.000, steps: 870
Episode 6: reward: 6.000, steps: 661
Episode 7: reward: 4.000, steps: 643
Episode 8: reward: 7.000, steps: 640
Episode 9: reward: 11.000, steps: 881
Episode 10: reward: 10.000, steps: 785
Recompensa media: 7.6


Como se puede ver, un mayor entrenamiento suele favorecer obtener una mayor recompensa, si bien no implica necesariamente que en cada episodio de test se vaya a obtener una mejor recompensa. Como se puede ver en la comparativa, hay casos con 60000 steps donde un episodio dio mejor resultado que con 490000 steps.

Los factores clave en la obtención de mejores resultados son:

*   El modelo escogido.
*   El valor de test en la política escogida.
*   La aleatoriedad del proceso, pues aunque se hagan más entrenamiento la aleatoriedad puede hacer que se obtengan peores resultados.
*   La duración del entrenamiento. A mayor número de entrenamientos, mayor probabilidad de obtener una mejor puntuación.



## Experimento 3

En la siguiente tabla, podemos observar los hiperparámetros utilizados para ciertas pruebas con la arquitectura que ofreció los mejores resultados en la evaluación, además de haberse podido completar el entrenamiento.

|  | Detalle | Prueba 1 | Prueba 2 | Prueba 3| Prueba 4|
|----------|----------|----------|----------|----------|----------|
| **RED** |   | 3 Conv2D   | 3 Conv2D| 3 Conv2D| 3 Conv2D|
|  |   | 1 Flatten  | 1 Flatten| 1 Flatten| 1 Flatten|
| |   | 3 Dense   | 3 Dense| 3 Dense| 3 Dense|
|   |  |    |   |   |    |
| **Parámetros** |  |  | | | |
| **Epsilon Greedy** | nb steps  | 10000   | 10000   | 10000   | 50000   |
| **Memory**   | limit   | 1000   | 1000   | 1000   | 50000   |
| **DQN Agent**   | nb steps warmup   | 10000  | 10000  | 10000  | 50000  |
|   | gamma   | 0.99   | 0.99   | 0.99   | 0.9   |
|   | target model update   | 10000   | 10000   | 5000   | 10000   |
| **DQN Compile**   | LR   | 0.0001   | 0.0001   | 0.0001   | 0.0001   |
| **DQN Fit**   | nb steps   | 250000   | 50000   | 10000   | 250000   |
|   |  |    |   |   |    |
| **TRAINING**   | Episodios   | 60   | 328   | 11   | 334   |
|   | Recompensa promedio  | 12.63   | 11.04   | 11.27   | 11.7   |
|   |  |    |   |   |    |
| **TEST**   | Episodios  | 10   | 10   | 10  | 10   |
|   | Recompensa promedio  | 8.4   | 10.4   | 13.5   | 12.1   |


Comparando las cuatro pruebas, observamos cómo los ajustes en los hiperparámetros impactaron tanto el entrenamiento como el test.

En Pruebas 1, 2 y 3, los valores de Epsilon Greedy (nb_steps=10,000) y la memoria limitada a 1000 restringieron la capacidad del agente para retener experiencias diversas. Esto resulta en una recompensa media en test de 8.4, 10.4 y 13.5, respectivamente. La Prueba 3 destaca porque, a pesar de entrenar solo 11 episodios (debido a nb_steps=10,000), obtuvo el mejor desempeño en test (13.5). Esto sugiere que el modelo pudo aprender rápidamente con una mayor frecuencia de actualización (target_model_update=5000), aunque el entrenamiento fue limitado en tiempo.

Por otro lado, en la Prueba 4, al incrementar la memoria a 50,000 y extender el período de exploración (Epsilon Greedy con nb_steps=50,000), el agente logró una recompensa más estable de 12.1 en test. Además, el nb_steps_warmup=50,000 proporcionó al modelo más tiempo para llenar su memoria antes de entrenar, permitiéndole aprender de una mayor cantidad de experiencias.

En conclusión, el mejor rendimiento en test se observó en la Prueba 3 debido a la mayor frecuencia de actualización del modelo objetivo y a un entrenamiento más eficiente, pues con pocos pasos ofreció mejores resultados.




# Conclusión


En este proyecto, se implementó y evaluó un agente Deep Q-Network (DQN) en el entorno SpaceInvaders-v0 con el objetivo de entrenarlo para alcanzar un desempeño competitivo. A través de diversas pruebas, se pudo observar que el agente fue capaz de aprender y mejorar su desempeño progresivamente, llegando finalmente a obtener una puntuación media de 22.3 en modo de prueba, a pesar de las limitaciones encontradas por google colab o en las ejecuciones en local, donde en prácticamente todos los intentos llegaba un punto en que dejaba de ejecutarse el entrenamiento antes de completarse.

El desarrollo del proyecto permitió comprender la importancia de un equilibrio entre exploración y explotación en el aprendizaje por refuerzo, así como la necesidad de tiempo y recursos adecuados para que el agente pueda alcanzar su máximo potencial.

En conclusión, este trabajo ha demostrado el potencial de los algoritmos de aprendizaje profundo aplicados al aprendizaje por refuerzo en entornos complejos. A pesar de las limitaciones, el agente logró un progreso significativo, sentando una base sólida para futuras mejoras y experimentos.